ethdb/pebble/pebble.go

```diff
@@ -0,0 +1,372 @@
+
+	// minHandles is the minimum number of files handles to allocate to the open
+	// database files.
+	minHandles = 16
+
+	// metricsGatheringInterval specifies the interval to retrieve leveldb database
+	// compaction, io and pause stats to report to the user.
+	metricsGatheringInterval = 3 * time.Second
+)
+
+// Database is a persistent key-value store based on the pebble storage engine.
+// Apart from basic data storage functionality it also supports batch writes and
+// iterating over the keyspace in binary-alphabetical order.
+type Database struct {
+	fn string     // filename for reporting
+	db *pebble.DB // Underlying pebble storage engine
+
+	compTimeMeter      metrics.Meter // Meter for measuring the total time spent in database compaction
+	compReadMeter      metrics.Meter // Meter for measuring the data read during compaction
+	compWriteMeter     metrics.Meter // Meter for measuring the data written during compaction
+	writeDelayNMeter   metrics.Meter // Meter for measuring the write delay number due to database compaction
+	writeDelayMeter    metrics.Meter // Meter for measuring the write delay duration due to database compaction
+	diskSizeGauge      metrics.Gauge // Gauge for tracking the size of all the levels in the database
+	diskReadMeter      metrics.Meter // Meter for measuring the effective amount of data read
+	diskWriteMeter     metrics.Meter // Meter for measuring the effective amount of data written
+	memCompGauge       metrics.Gauge // Gauge for tracking the number of memory compaction
+	level0CompGauge    metrics.Gauge // Gauge for tracking the number of table compaction in level0
+	nonlevel0CompGauge metrics.Gauge // Gauge for tracking the number of table compaction in non0 level
+	seekCompGauge      metrics.Gauge // Gauge for tracking the number of table compaction caused by read opt
+
+	quitLock sync.Mutex      // Mutex protecting the quit channel access
+	quitChan chan chan error // Quit channel to stop the metrics collection before closing the database
+
+	log log.Logger // Contextual logger tracking the database path
+}
+
+// New returns a wrapped LevelDB object. The namespace is the prefix that the
+// metrics reporting should use for surfacing internal stats.
+func New(file string, cache int, handles int, namespace string) (*Database, error) {
+	// Ensure we have some minimal caching and file guarantees
+	if cache < minCache {
+		cache = minCache
+	}
+	if handles < minHandles {
+		handles = minHandles
+	}
+	logger := log.New("database", file)
+	logger.Info("Allocated cache and file handles", "cache", common.StorageSize(cache*1024*1024), "handles", handles)
+
+	// Open the db and recover any potential corruptions
+	db, err := pebble.Open(file, &pebble.Options{
+		// SStable and WALs' sync rate. The default rate is 512KB.
+		// It's worthwhile to tune the performance by this setting.
+		BytesPerSync: 0, // A larger number maybe is better?
+		Cache:        pebble.NewCache(int64(cache * 1024 * 1024 / 2)),
```

Comment:
This was the only way I could force all levels to be 2MB. This may or may not be a good idea, mostly just trying to emulate whatever leveldb is doing so we have comparable numbers:

```
		// Per-level options. Options for at least one level must be specified. The
		// options for the last level are used for all subsequent levels.
		Levels: []pebble.LevelOptions{
			{TargetFileSize: 2 * 1024 * 1024},
			{TargetFileSize: 2 * 1024 * 1024},
			{TargetFileSize: 2 * 1024 * 1024},
			{TargetFileSize: 2 * 1024 * 1024},
			{TargetFileSize: 2 * 1024 * 1024},
			{TargetFileSize: 2 * 1024 * 1024},
			{TargetFileSize: 2 * 1024 * 1024},
		},
```

Reply:
We don't need the `/2` in Pebbel. LevelDB defines the read and write caches separately. Pebble has a single combined cache area and the write buffers are taken from this too.